lib/ui/painting/image_decoder.cc

```diff
@@ -122,31 +144,102 @@ static sk_sp<SkImage> ImageFromDecompressedData(
   flow.Step(__FUNCTION__);
 
-  auto decoded_image = SkImage::MakeFromEncoded(data);
+  if (!target_width && !target_height) {
+    // The resized dimesions are the same as the intrinsic dimensions of the
+    // image. There is nothing to do.
+    return SkImage::MakeFromEncoded(data)->makeRasterImage();
+  }
 
-  if (!decoded_image) {
+  auto codec = SkCodec::MakeFromData(data);
+  FML_DCHECK(codec);
+  if (codec == nullptr) {
+    FML_LOG(ERROR) << "Failed to instantiate image decoder.";
     return nullptr;
   }
 
-  // Make sure to resolve all lazy images.
-  decoded_image = decoded_image->makeRasterImage();
+  const auto* codec_ptr = codec.get();
+
+  // Note that we cannot read the dimensions from the codec since they don't
+  // respect image orientation provided e.g. in EXIF data.
+  auto image_generator = SkCodecImageGenerator::MakeFromCodec(std::move(codec));
+  const auto& source_dimensions = image_generator->getInfo().dimensions();
+
+  auto resized_dimensions =
+      GetResizedDimensions(source_dimensions, target_width, target_height);
+
+  // No resize needed.
+  if (resized_dimensions == source_dimensions) {
+    return SkImage::MakeFromEncoded(data)->makeRasterImage();
+  }
+
+  auto decode_dimensions = codec_ptr->getScaledDimensions(
+      std::max(static_cast<double>(resized_dimensions.width()) /
+                   source_dimensions.width(),
+               static_cast<double>(resized_dimensions.height()) /
+                   source_dimensions.height()));
+
+  // If the codec supports efficient sub-pixel decoding, decoded at a resolution
+  // close to the target resolution before resizing.
+  if (decode_dimensions != codec_ptr->dimensions()) {
+    if (source_dimensions != codec_ptr->dimensions()) {
+      decode_dimensions =
+          SkISize::Make(decode_dimensions.height(), decode_dimensions.width());
+    }
+
+    auto scaled_image_info =
+        image_generator->getInfo().makeDimensions(decode_dimensions);
+
+    SkBitmap scaled_bitmap;
+    if (!scaled_bitmap.tryAllocPixels(scaled_image_info)) {
+      FML_LOG(ERROR) << "Could not allocate bitmap when attempting to scale.";
+      return nullptr;
+    }
+
+    const auto& pixmap = scaled_bitmap.pixmap();
+    if (image_generator->getPixels(pixmap.info(), pixmap.writable_addr(),
+                                   pixmap.rowBytes())) {
+      // Marking this as immutable makes the MakeFromBitmap call share
+      // the pixels instead of copying.
+      scaled_bitmap.setImmutable();
+
+      auto decoded_image = SkImage::MakeFromBitmap(scaled_bitmap);
+      FML_DCHECK(decoded_image);
+      if (!decoded_image) {
+        FML_LOG(ERROR)
+            << "Could not create a scaled image from a scaled bitmap.";
+        return nullptr;
+      }
+      return ResizeRasterImage(std::move(decoded_image), resized_dimensions,
+                               flow);
+    }
+  }
 
-  if (!decoded_image) {
+  auto image = SkImage::MakeFromEncoded(data);
+  FML_DCHECK(image);
```

Comment:
This is not a particularly exceptional case and tripping an assertion to terminate the process is too drastic.

Reply:
This should never happen.

How could we get here from where we are? We already successfully allocated the pixels, we already have a valid codec.  Something's seriously wrong if the image is null here.

Reply:
> This should never happen.

Note this is SkImage::MakeFromEncoded, so it would also happen if you couldn't find a codec, etc.

Removed the assertion.